### DISCUSSION TITLE
Add particles and bonds after simulation is initialized

### DISCUSSION THREAD

**Discussion by yliu6677** [glotzerlab/hoomd-blue#1858] [Support] [answered]
What is the best way to add particles and bonds after simulation is initialized?

At the moment I have initialized and minimized my two particle system with a spring connecting them, I want to add a third particle and form two new spring to form a triangle and minimize, while recording the potential energy and gsd frame.

```
import hoomd
import math
import numpy as np
import itertools
import gsd.hoomd
import hoomd.md.integrate
import h5py

cpu = hoomd.device.CPU()
simulation = hoomd.Simulation(device=cpu)

run_time = 500
record_time = 500

Box_L = 10
N_particles = 2
spacing = 5

x = np.linspace(-spacing, spacing, N_particles, endpoint=False)

init_pos = list(itertools.product(x, repeat=3))[0: N_particles]
orientation = [(1, 0, 0, 0)] * N_particles

frame = gsd.hoomd.Frame()
frame.particles.N = N_particles
frame.particles.position = [[-1, -1, -1], [1, 1, 0.3]]
frame.particles.orientation = orientation
frame.particles.typeid = [0] * N_particles
frame.particles.types = ["A"]
frame.configuration.box = [Box_L, Box_L, Box_L, 0, 0, 0]

harmonic = hoomd.md.bond.Harmonic()
harmonic.params['A-A'] = dict(k=3.0, r0=2.76)
frame.bonds.N = 1
frame.bonds.types = ['A-A']
frame.bonds.typeid = [0] * 1
frame.bonds.group = [[0, 1]]

simulation.create_state_from_snapshot(frame)
hoomd.write.GSD.write(state=simulation.state, filename='frame.gsd', mode='xb')

constant_volume = hoomd.md.methods.ConstantVolume(filter=hoomd.filter.All())
fire = hoomd.md.minimize.FIRE(
    dt=0.001, force_tol=1e-3, angmom_tol=1e-3, energy_tol=1e-6, forces=[harmonic], methods=[constant_volume]
)
simulation.operations.integrator = fire
simulation.run(0)

thermodynamic_properties = hoomd.md.compute.ThermodynamicQuantities(
    filter=hoomd.filter.All()
)
simulation.operations.computes.append(thermodynamic_properties)

logger = hoomd.logging.Logger(categories=['scalar'])
logger.add(thermodynamic_properties, quantities="potential_energy")

hdf5_writer = hoomd.write.HDF5Log(
    trigger=hoomd.trigger.Periodic(record_time), filename='log.h5', mode='a', logger=logger
)
simulation.operations.writers.append(hdf5_writer)

gsd_writer = hoomd.write.GSD(
    filename='frame.gsd', trigger=hoomd.trigger.Periodic(record_time), mode='ab'
)
simulation.operations.writers.append(gsd_writer)

while not fire.converged:
    simulation.run(run_time)
```

**Comment by joaander** [comment]
The simplest solution is to place the first two particles at a distance equal to the the minimum of the bond potential, then add the third before calling `simulation.create_state_from_snapshot`.

[State.set_snapshot](https://hoomd-blue.readthedocs.io/en/v4.8.2/package-hoomd.html#hoomd.State.set_snapshot) is the only Python API method that allows you to change the number of particles/bonds/etc...

**Reply by yliu6677** [reply]
Can you say how to change the # of particles/bonds/etc by `State.set_snapshot(snap)'? I went through it and found it might be used to restore a state to several runs before with exactly the same system. Errors will be raised if the system does not match.

**Reply by joaander** [reply]
[The documentation](https://hoomd-blue.readthedocs.io/en/v4.8.2/package-hoomd.html#hoomd.State.set_snapshot) states:
> [set_snapshot](https://hoomd-blue.readthedocs.io/en/v4.8.2/package-hoomd.html#hoomd.State.set_snapshot) can only make limited changes to the simulation state. While it can change the number of particles/bonds/etc… or their properties, it cannot change the number or names of the particle/bond/etc.. types.

**Reply by yliu6677** [reply]
However I did not find an example of adding particles using `set_snapshot`. My way below is more flexible since one can append or alter anything to a `gsd.hoomd.Frame()`.

**Comment by yliu6677** [comment, marked as answer]
Here is my solution: 

Create a gsd frame by 'frame = gsd.hoomd.Frame()'. 

Copy the current snapshot to this frame.

Add particles etc to this frame.

Create a new simulation with the state from this frame.

Please leave your solution if you have a better one.
